mining/src/manager.rs

```diff
@@ -0,0 +1,178 @@
+                    }
+                }
+                Err(err) => {
+                    return Err(err)?;
+                }
+            }
+        }
+    }
+
+    pub(crate) fn block_candidate_transactions(&self) -> Vec<MutableTransaction> {
+        self.mempool.read().block_candidate_transactions()
+    }
+
+    /// Clears the block template cache, forcing the next call to get_block_template to build a new block template.
+    pub fn clear_block_template(&self) {
+        self.block_template_cache.write().clear();
+    }
+
+    #[cfg(test)]
+    pub(crate) fn block_template_builder(&self) -> &BlockTemplateBuilder {
+        &self.block_template_builder
+    }
+
+    /// validate_and_insert_transaction validates the given transaction, and
+    /// adds it to the set of known transactions that have not yet been
+    /// added to any block.
+    ///
+    /// The returned transactions are clones of objects owned by the mempool.
+    pub fn validate_and_insert_transaction(
+        &mut self,
+        transaction: MutableTransaction,
+        is_high_priority: bool,
+        allow_orphan: bool,
+    ) -> MiningManagerResult<Vec<MutableTransaction>> {
+        Ok(self.mempool.write().validate_and_insert_transaction(transaction, is_high_priority, allow_orphan)?)
+    }
+
+    /// Try to return a mempool transaction by its id.
+    ///
+    /// Note: the transaction is an orphan if tx.is_fully_populated() returns false.
+    pub fn get_transaction(
+        &self,
+        transaction_id: &TransactionId,
+        include_transaction_pool: bool,
+        include_orphan_pool: bool,
+    ) -> Option<MutableTransaction> {
+        self.mempool.read().get_transaction(transaction_id, include_transaction_pool, include_orphan_pool)
+    }
+
+    pub fn get_all_transactions(
+        &self,
+        include_transaction_pool: bool,
+        include_orphan_pool: bool,
+    ) -> (Vec<MutableTransaction>, Vec<MutableTransaction>) {
+        self.mempool.read().get_all_transactions(include_transaction_pool, include_orphan_pool)
+    }
+
+    /// get_transactions_by_addresses returns the sending and receiving transactions for
+    /// a set of addresses.
+    ///
+    /// Note: a transaction is an orphan if tx.is_fully_populated() returns false.
+    pub fn get_transactions_by_addresses(
+        &self,
+        script_public_keys: &ScriptPublicKeySet,
+        include_transaction_pool: bool,
+        include_orphan_pool: bool,
+    ) -> OwnerSetTransactions {
+        self.mempool.read().get_transactions_by_addresses(script_public_keys, include_transaction_pool, include_orphan_pool)
+    }
+
+    pub fn transaction_count(&self, include_transaction_pool: bool, include_orphan_pool: bool) -> usize {
+        self.mempool.read().transaction_count(include_transaction_pool, include_orphan_pool)
+    }
+
+    pub fn handle_new_block_transactions(&self, block_transactions: &[Transaction]) -> MiningManagerResult<Vec<MutableTransaction>> {
```

Comment:
Add a TODO about using tx acceptance data to verify that new block txs are actually accepted into virtual state. On the broader level this is related to mempool replay feature well